### ISSUE TITLE
Support for unlimited precision floating point

### ISSUE THREAD

**Issue by nomennescio** [factor/factor#2636]
Would it make sense to have an "unlimited" precision floating point library?
It could be build on top of bignum primitives. I think that would be quite a unique feature.
To pull it off, there would have to be an (arbitrary) cutoff (which would make it almost like a `ratio`).

**Comment by bpollack** [comment]
Are you describing fixed-point arithmetic? If so, I don't quite get what the difference would be here versus what `ratio`, since any fixed-point number could be represented as `n+m/10^p`. I can see wanting a quick way to enter fixed-point as a literal and print it as a precision-preserving decimal, though; that wouldn't be too hard to add, I wouldn't think.

**Comment by nomennescio** [comment]
No, because it still has an exponent.

**Comment by jonenst** [comment]
Are you talking about more efficiently representing (in number of bytes) some numbers ?

for example 1.23e-500 is stored as a ratio with (guessing): ~200bytes ? 123 (fixnum, 8 bytes) divided by 1000...00000 (bignum with 500 zeros: 200bytes?)
But could be stored with exponents in 20 bytes ?

Of course all of this is about tradeoffs, if you have less bytes for some numbers, you'll have more bytes for others.

**Comment by jonenst** [comment]
Or having more performant operations ?
Like 123e-500 * 20e-200 is fast to compute with exponents ans slower with ratio

**Comment by nomennescio** [comment]
Good questions! I think it's mainly about accuracy; sooner or later with 32- or 64-bit floats, things will get inaccurate, by extending the number of bits used for floating point, you can do more accurate calculations, similar how bignum does for integer calculations. Of course, any irrational number would need an infinite number of digits to be represented, which is not possibly, so some kind of cut-off has to be defined somehow. But once numbers have a limited numbers of bits, you can calculate the number of bits needed to not loose accuracy and use that.

**Comment by jonenst** [comment]
But if it's mainly about accuracy, how are they different from ratio ? What do you mean by
"No, because it would still have an exponent."
What's the difference between these
```
ratio:
1231231232309482304982039482304912312/10000000000000000000000000000000000000000

"infinite precision floating point":
123.1231232309482304982039482304912312

"infinite precision floating point with explicit exponent":
1231231232309482304982039482304912312e-50
```
if it's not the number of bytes to store them, or the performance of operations, I can't see how else you would differentiate between the two. The default string representation of them ? So maybe just add a word to display ratios as decimals? In these cases, I mostly use `>float` or `>double` because when I want the string representation I usually do not care about the infinite precision, although I see how it may be useful.

Writing this reminded me about https://docs.factorcode.org/content/vocab-decimals.html which seems to be what you are talking about in base 10 ? There is no ratio>decimal word though, but I think I wrote one in the past. Writing ratio>decimal is easy, you just have to check if the prime factors of the denominator (in the reduced form, which factor ratios always are I think) are only 2s and 5s to know if it's finite or infinite, and decide if you throw an error for ratios with infinite decimal representations (e.g 1/3 = 0.3333 can never be represented exactly by a DECIMAL: by nature) or allow the user to decide the precision (1/3 25 ratio>decimal -> DECIMAL: 0.33333333333333, 2/3 ratio>decimal -> DECIMAL: 0.66666666667)

so the decimals vocabulary gives you
- an exact precision string representation for finite ratio
- an arbitrary length precision for infinite ratio 

I think you could even (in a separate vocabulary I guess) have exact precision for all ratios (especially infinite ones) in some kind of extended decimal form where you mark the repeating pattern ? https://en.wikipedia.org/wiki/Repeating_decimal
> A more complicated example is 3227/555, whose decimal becomes periodic at the second digit following the decimal point and then repeats the sequence "144" forever, i.e. 5.8144144144.... At present, there is no single universally accepted [notation or phrasing](https://en.wikipedia.org/wiki/Repeating_decimal#Notation) for repeating decimals. From the representations in the wikipedia table (  Vinculum | Dots | Parentheses | Arc | Ellipsis ) parentheses seems to be the one that's best suited for programming languages (it looks bijective, whereas ellipsis is not ? others would use weird unicode chars)

**Comment by nomennescio** [comment]
In a sense all floating point numbers are "just ratios" with denominator a power of two, yet they're useful on their own.  When calculating functions such as `sin` on ratios, they're converted to float first with '>float', and then lose accuracy.  Having a floating point type with "unlimited" precision could keep the accuracy. But you raise important questions. I will think a bit about it. I will also have a look at the `decimal` vocab. Your point about repeating digits is interesting, but would lead to a separate numeric class. Similarly a separate numeric class for the algebraic numbers could be made to accurately represent them. That's outside the scope of this proposal however.

**Comment by jonenst** [comment]
For sin and all other functions, the only thing you can do I think is to keep everything symbolic, and then when you want do display actual digits to an arbitrary precision, compute the whole thing with iterative approximation formulas. I don't think that's an objective for a general purpose programming language like factor, it's more in the domain of math specific languages like mathematica or wolframalpha ?

**Comment by mrjbq7** [comment]
I was wondering if you wanted to go in the direction of a N-bit precision float library code so you could make 96-bit floats, or 256-bit floats, and have the library perform all the float typical computations with exponents/mantissa.

**Comment by nomennescio** [comment]
> I was wondering if you wanted to go in the direction of a N-bit precision float library code so you could make 96-bit floats, or 256-bit floats, and have the library perform all the float typical computations with exponents/mantissa.

Yes, with automatic promotions to keep accuracy (and maybe operations to limit precision), just as bignums keep accuracy.

**Comment by nomennescio** [comment]
> Writing this reminded me about https://docs.factorcode.org/content/vocab-decimals.html which seems to be what you are talking about in base 10 ? There is no ratio>decimal word though, but I think I wrote one in the past. Writing ratio>decimal is easy, you just have to check if the prime factors of the denominator (in the reduced form, which factor ratios always are I think) are only 2s and 5s to know if it's finite or infinite, and decide if you throw an error for ratios with infinite decimal representations (e.g 1/3 = 0.3333 can never be represented exactly by a DECIMAL: by nature) or allow the user to decide the precision (1/3 25 ratio>decimal -> DECIMAL: 0.33333333333333, 2/3 ratio>decimal -> DECIMAL: 0.66666666667)
> 
> so the decimals vocabulary gives you
> 
> * an exact precision string representation for finite ratio
> * an arbitrary length precision for infinite ratio
> 

I did look at it, and it indeed can represent `ratio * 10^exponent`, but lacks the support of using it with irrational functions such as 'sin', and is not integrated in the the `MATH:` hierarchy of classes. For true floating-point, it makes more sense to have `ratio*2^exponent`, and to support functions on them. At some point you would need to gracefully handle cutoffs, e.g. when using Maclaurin approximations. One way would be to have each number carry an explicit or implicit `epsilon` and do all calculations  based on that accuracy.
